Dockerfile

```diff
@@ -0,0 +1,43 @@
+FROM ubuntu:bionic as source
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
+    git ca-certificates
+
+COPY . selfoss/
+
+FROM composer:1.9 as composer
+
+COPY --from=source /selfoss /selfoss
+
+RUN cd /selfoss && composer install --ignore-platform-reqs --optimize-autoloader --no-dev
+
+
+
+FROM node:13-buster as npm
+
+COPY --from=composer /selfoss /selfoss
+
+RUN cd /selfoss && npm i && npm install --prefix assets/ && npm run build
+
+
+
+FROM php:7.4-apache-buster
+
+COPY --from=npm /selfoss /var/www/html/
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y libpng-dev \
+    && docker-php-ext-install -j$(nproc) gd \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /var/www/html/
+
+RUN mkdir config && ln -s config/config.ini config.ini \
+    && a2enmod rewrite \
+    && /bin/echo -e "#!/bin/bash\nchown -R www-data:www-data /var/www/html/data/cache /var/www/html/data/favicons /var/www/html/data/logs /var/www/html/data/thumbnails /var/www/html/data/sqlite" > /entrypoint.sh \
```

Comment:
I find this ugly. Could we create `utils/docker/entrypoint.sh` and call that here?

Reply:
> I find this ugly. Could we create `utils/docker/entrypoint.sh` and call that here?

Yes, it's a good idea but unfortunately, I'm not that good with bash/shell scripting :(

Reply:
I think just literally calling the command should work:

```
mkdir utils/docker \
&& echo -e "#!/bin/bash\nchown -R www-data:www-data /var/www/html/data/cache /var/www/html/data/favicons /var/www/html/data/logs /var/www/html/data/thumbnails /var/www/html/data/sqlite" > utils/docker/entrypoint.sh
&& echo -e "if [ ! -f config/config.ini ]; then cp defaults.ini config/config.ini && sed -i 's/logger_destination=.*$/logger_destination=file:php:\/\/stderr/' config/config.ini; fi" >> utils/docker/entrypoint.sh \
&& echo -e "su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php' >/dev/null 2>&1" >> utils/docker/entrypoint.sh \
&& echo -e "(while true; do su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php'; sleep 900; done;) &" >> utils/docker/entrypoint.sh \
&& echo -e "apache2-foreground" >> utils/docker/entrypoint.sh \
&& chmod a+x utils/docker/entrypoint.sh
```

Plus maybe some clean up after that.

Reply:
> I think just literally calling the command should work:
> 
> ```
> mkdir utils/docker \
> && echo -e "#!/bin/bash\nchown -R www-data:www-data /var/www/html/data/cache /var/www/html/data/favicons /var/www/html/data/logs /var/www/html/data/thumbnails /var/www/html/data/sqlite" > utils/docker/entrypoint.sh
> && echo -e "if [ ! -f config/config.ini ]; then cp defaults.ini config/config.ini && sed -i 's/logger_destination=.*$/logger_destination=file:php:\/\/stderr/' config/config.ini; fi" >> utils/docker/entrypoint.sh \
> && echo -e "su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php' >/dev/null 2>&1" >> utils/docker/entrypoint.sh \
> && echo -e "(while true; do su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php'; sleep 900; done;) &" >> utils/docker/entrypoint.sh \
> && echo -e "apache2-foreground" >> utils/docker/entrypoint.sh \
> && chmod a+x utils/docker/entrypoint.sh
> ```
> 
> Plus maybe some clean up after that.

Where should I call this command? in Dockerfile or on my local system to create the entrypoint.sh?

Reply:
On the local system. It should create the file on your local system and you can then just call `utils/docker/entrypoint.sh` in the `Dockerfile`.

Reply:
> On the local system. It should create the file on your local system and you can then just call `utils/docker/entrypoint.sh` in the `Dockerfile`.

Ok, going to try that.

Reply:
> On the local system. It should create the file on your local system and you can then just call `utils/docker/entrypoint.sh` in the `Dockerfile`.

Ok, yes, the command works and the file is created.
can you please check it?
If possible then can you please add the required cleanups which you were talking about?
(I've changed the extention for now because github wasn't allowing the .sh uploads)

Reply:
This looks more usable:

```bash
#!/bin/bash

# Make data directories owned by Apache
chown -R www-data:www-data \
    /var/www/html/data/cache \
    /var/www/html/data/favicons \
    /var/www/html/data/logs \
    /var/www/html/data/thumbnails \
    /var/www/html/data/sqlite

# Create a config file when one does not exist
if [[ ! -f config/config.ini ]]; then
    cp defaults.ini config/config.ini \
    && sed -i 's#^logger_destination=.*#logger_destination=file:php://stderr#' config/config.ini
fi

# Run updater process periodically
su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php' >/dev/null 2>&1
(while true; do su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php'; sleep 900; done;) &

# Start the server
apache2-foreground
```

Though I am not sure about several things:
* Does the owner need to be changed? How does that translate across the mounted volume? What if www-data (or user with the same uid) does not exist outside of Docker container?
* Do we need to create the config file when it does not exists? selfoss will happily read the data from `defaults.ini` when some keys or even the whole `config.ini` is missing.
    * Maybe discoverability is a boon here, since `defaults.ini` will be hidden inside the container.
* Do we need to set the logger destination in the config, or might it make better sense forcing it through `SELFOSS_LOGGER_DESTINATION` environment variable.
* The while loop inside subprocess is very ugly. Does docker have some mechanism for running commands periodically like cron or systemd timers? I found https://developers.redhat.com/blog/2019/04/24/how-to-run-systemd-in-a-container/ but do not have time to study it in detail.

Reply:
> This looks more usable:
> 
> ```shell
> #!/bin/bash
> 
> # Make data directories owned by Apache
> chown -R www-data:www-data \
>     /var/www/html/data/cache \
>     /var/www/html/data/favicons \
>     /var/www/html/data/logs \
>     /var/www/html/data/thumbnails \
>     /var/www/html/data/sqlite
> 
> # Create a config file when one does not exist
> if [[ ! -f config/config.ini ]]; then
>     cp defaults.ini config/config.ini \
>     && sed -i 's#^logger_destination=.*#logger_destination=file:php://stderr#' config/config.ini
> fi
> 
> # Run updater process periodically
> su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php' >/dev/null 2>&1
> (while true; do su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php'; sleep 900; done;) &
> 
> # Start the server
> apache2-foreground
> ```

> Ok, I'll create the entrypoint.sh from this text.


> 
> Though I am not sure about several things:
> 

> > * Does the owner need to be changed? How does that translate across the mounted volume? What if www-data (or user with the same uid) does not exist outside of Docker container?

> Not sure about it, @squatica can you please help here?



> * Do we need to create the config file when it does not exists? selfoss will happily read the data from `defaults.ini` when some keys or even the whole `config.ini` is missing.
>   
> I think no, but I don't know why, it is still not taking any configurations from config.ini while running on docker, I think I'll have to investigate more on this.

>   * Maybe discoverability is a boon here, since `defaults.ini` will be hidden inside the container.
> * Do we need to set the logger destination in the config, or might it make better sense forcing it through `SELFOSS_LOGGER_DESTINATION` environment variable.

> We can any of the option from above, what do you recommend? should we use environment variables?

> * The while loop inside subprocess is very ugly. Does docker have some mechanism for running commands periodically like cron or systemd timers? I found https://developers.redhat.com/blog/2019/04/24/how-to-run-systemd-in-a-container/ but do not have time to study it in detail.


> Actually we'll have to install crontab inside the container base image, I'll take a look at it soon.

> This looks more usable:
> 
> ```shell
> #!/bin/bash
> 
> # Make data directories owned by Apache
> chown -R www-data:www-data \
>     /var/www/html/data/cache \
>     /var/www/html/data/favicons \
>     /var/www/html/data/logs \
>     /var/www/html/data/thumbnails \
>     /var/www/html/data/sqlite
> 
> # Create a config file when one does not exist
> if [[ ! -f config/config.ini ]]; then
>     cp defaults.ini config/config.ini \
>     && sed -i 's#^logger_destination=.*#logger_destination=file:php://stderr#' config/config.ini
> fi
> 
> # Run updater process periodically
> su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php' >/dev/null 2>&1
> (while true; do su www-data -s /bin/bash -c 'php /var/www/html/cliupdate.php'; sleep 900; done;) &
> 
> # Start the server
> apache2-foreground
> ```
> 
> Though I am not sure about several things:
> 
> * Does the owner need to be changed? How does that translate across the mounted volume? What if www-data (or user with the same uid) does not exist outside of Docker container?
> * Do we need to create the config file when it does not exists? selfoss will happily read the data from `defaults.ini` when some keys or even the whole `config.ini` is missing.
>   
>   * Maybe discoverability is a boon here, since `defaults.ini` will be hidden inside the container.
> * Do we need to set the logger destination in the config, or might it make better sense forcing it through `SELFOSS_LOGGER_DESTINATION` environment variable.
> * The while loop inside subprocess is very ugly. Does docker have some mechanism for running commands periodically like cron or systemd timers? I found https://developers.redhat.com/blog/2019/04/24/how-to-run-systemd-in-a-container/ but do not have time to study it in detail.

Reply:
Ooof, I see you've been really busy lately. Let me catch up.

> Does the owner need to be changed? How does that translate across the mounted volume?

The user inside the container is root, but apache2-foreground automatically forks and then runs under www-data (this user is created during apache installation) so that's why we also have to run the cliupdate.php under www-data otherwise the first call to cliupdate would create the sqlite database with root permissions and then www-data would not be able to write in it.

>  What if www-data (or user with the same uid) does not exist outside of Docker container?

It doesn't have to exist outside the docker container. As I mentioned in another comment, this image was meant to be the production type with sources from git, and in such case we will never need to match the user inside with the user outside. Outside user will not need to write into these volumes (if so, they can do it by docker exec into this container and then they have full rights). The volume doesn't even need to be mounted to a directory in the host filesystem, docker will manage it somewhere in the background in /var/lib/docker or so.

Reply:
> Do we need to create the config file when it does not exists? selfoss will happily read the data from defaults.ini when some keys or even the whole config.ini is missing.

This was actually something I had to work around - I had to create a separate directory for config and then symlink the config.ini into the right place. Because volumes can only be directories, not single files. Usually it's better to have user-changeable things such as configs in one directory, separate from the source files. And as mentioned previously, I'm checking for the existence of config.ini because we expect the volume to be reused, but we also want to initialize it when the volume is empty in the first run.

Reply:
> Do we need to set the logger destination in the config, or might it make better sense forcing it through SELFOSS_LOGGER_DESTINATION environment variable.

I wanted to have the selfoss logs visible in `docker logs` of the whole container (and therefore also in `docker-compose logs` which is the first place to check if something doesn't work as expected). This is where docker includes the output of the CMD command, our CMD is apache2-foreground so `docker logs` will show access logs, but that's not so useful. In the meantime selfoss logs might contain important errors and they would be hidden un-intuitively in data/logs. By switching to `file:php://stderr` selfoss errors are included among the apache access logs. I'm not sure what SELFOSS_LOGGER_DESTINATION does - maybe you can achieve the same behavior with it?

Reply:
> The while loop inside subprocess is very ugly.

It's actually quite normal thing in docker containers. Running two main processes (cron + apache) is discouraged - each container should only run one process. Same problem with systemd (plus a bunch of other permission related problems). While loop is simple and efficient and it will die when the container dies, so no problem. If we wanted to be completely pure about it, we would have to include something like supervisorctl and run apache and this cliupdate loop under supervisor. But that's too much hassle just for two processes. Alternative is to call the cliupdate from outside by `docker exec <container> php /var/www/html/cliupdate.php` but then we would force the user to do some extra setup outside of just docker-compose up.

Reply:
About the entrypoint.sh - yes, thanks for putting it into utils, the only reason why I used echo in Dockerfile was to have two Dockerfiles - one for 2.18 and one for master and I didn't want to have two entrypoint.sh files (they differ in the chown lines).

Reply:
> Seems the right approach to me. @jtojnar opinions?

> > Do we need to set the logger destination in the config, or might it make better sense forcing it through SELFOSS_LOGGER_DESTINATION environment variable.
> 
> I wanted to have the selfoss logs visible in `docker logs` of the whole container (and therefore also in `docker-compose logs` which is the first place to check if something doesn't work as expected). This is where docker includes the output of the CMD command, our CMD is apache2-foreground so `docker logs` will show access logs, but that's not so useful. In the meantime selfoss logs might contain important errors and they would be hidden un-intuitively in data/logs. By switching to `file:php://stderr` selfoss errors are included among the apache access logs. I'm not sure what SELFOSS_LOGGER_DESTINATION does - maybe you can achieve the same behavior with it?

Reply:
> can't we use cron? will it cause issue?
> > The while loop inside subprocess is very ugly.
> 
> It's actually quite normal thing in docker containers. Running two main processes (cron + apache) is discouraged - each container should only run one process. Same problem with systemd (plus a bunch of other permission related problems). While loop is simple and efficient and it will die when the container dies, so no problem. If we wanted to be completely pure about it, we would have to include something like supervisorctl and run apache and this cliupdate loop under supervisor. But that's too much hassle just for two processes. Alternative is to call the cliupdate from outside by `docker exec <container> php /var/www/html/cliupdate.php` but then we would force the user to do some extra setup outside of just docker-compose up.